node/network/statement-distribution/src/lib.rs

```diff
@@ -229,60 +249,63 @@ impl<R: rand::Rng> StatementDistributionSubsystem<R> {
+				StatementDistributionMessage::Share(relay_parent, statement) => {
+					let _timer = metrics.time_share();
 
-						// pass to legacy if legacy state contains head.
-						if legacy_v1_state.contains_relay_parent(&relay_parent) {
-							crate::legacy_v1::share_local_statement(
-								ctx,
-								legacy_v1_state,
-								relay_parent,
-								StatementWithPVD::drop_pvd_from_signed(statement),
-								&mut self.rng,
-								metrics,
-							)
+					// pass to legacy if legacy state contains head.
+					if legacy_v1_state.contains_relay_parent(&relay_parent) {
+						crate::legacy_v1::share_local_statement(
+							ctx,
+							legacy_v1_state,
+							relay_parent,
+							StatementWithPVD::drop_pvd_from_signed(statement),
+							&mut self.rng,
+							metrics,
+						)
+						.await?;
+					} else {
+						vstaging::share_local_statement(ctx, state, relay_parent, statement)
 							.await?;
-						}
-					},
-					StatementDistributionMessage::NetworkBridgeUpdate(event) => {
-						// pass to legacy, but not if the message isn't
-						// v1.
-						let legacy = match &event {
-							&NetworkBridgeEvent::PeerMessage(_, ref message) => match message {
-								Versioned::VStaging(protocol_vstaging::StatementDistributionMessage::V1Compatibility(_)) => true,
-								Versioned::V1(_) => true,
-								Versioned::VStaging(_) => false,
-							},
-							_ => true,
-						};
-
-						if legacy {
-							crate::legacy_v1::handle_network_update(
-								ctx,
-								legacy_v1_state,
-								v1_req_sender,
-								event,
-								&mut self.rng,
-								metrics,
-							)
-							.await;
-						}
+					}
+				},
+				StatementDistributionMessage::NetworkBridgeUpdate(event) => {
+					// pass to legacy, but not if the message isn't v1.
+					let (legacy, peer_id) = match &event {
+						&NetworkBridgeEvent::PeerMessage(peer_id, ref message) => match message {
+							Versioned::VStaging(
+								protocol_vstaging::StatementDistributionMessage::V1Compatibility(_),
+							) => (true, Some(peer_id)),
+							Versioned::V1(_) => (true, Some(peer_id)),
+							Versioned::VStaging(_) => (false, Some(peer_id)),
+						},
+						_ => (true, None),
+					};
 
-						// TODO [now]: pass to vstaging, but not if the message is
-						// v1 or the connecting peer is v1.
-					},
-					StatementDistributionMessage::Backed(candidate_hash) => {
-						crate::vstaging::handle_backed_candidate_message(
+					if legacy {
+						crate::legacy_v1::handle_network_update(
 							ctx,
-							unimplemented!(), // TODO [now] state
-							candidate_hash,
+							legacy_v1_state,
+							v1_req_sender,
+							event,
+							&mut self.rng,
+							metrics,
 						)
 						.await;
-					},
+					} else if peer_id
+						.map_or(false, |peer_id| !legacy_v1_state.peers.contains_key(&peer_id))
+					{
+						// pass to vstaging, but not if the message is
+						// v1 or the connecting peer is v1.
+						// TODO: Is the check above correct?
```

Comment:
Not sure about the check.

Reply:
I don't think we need the check. In fact, I think the check is wrong: nodes upgraded to the vstaging protocol should still use the `legacy_v1` protocol  until the runtime activates prospective parachains, and therefore be considered connected on `legacy_v1`. A simple `Else` would suffice, internal logic will handle any unexpected messages.